addons/package_hierarchy/models/package_links.py

```diff
@@ -0,0 +1,182 @@
+
+    def construct(self):
+        for parent_package, links in self.groupby("parent_id"):
+            links.mapped("child_id").write(
+                {"parent_id": parent_package.id if parent_package else False}
+            )
+
+    @api.model
+    def create_top_level_unlinks(self, packages, move_lines=False):
+        # Warm the cache
+        packages.get_move_lines(aux_domain=[("id", "in", move_lines.ids)])
+        link_vals = []
+        # Create unlinks from higher level packages
+        for top_level_pack in packages:
+            if top_level_pack.parent_id:
+                mls = top_level_pack.get_move_lines(aux_domain=[("id", "in", move_lines.ids)])
+                link_vals.append(
+                    {
+                        "parent_id": False,
+                        "child_id": top_level_pack.id,
+                        "move_line_ids": [(6, 0, mls.ids if mls else False)],
+                    }
+                )
+
+        if link_vals:
+            return self.create(link_vals)
+        return self.browse()
+
+    @api.depends("parent_id", "child_id")
+    def _compute_name(self):
+        for link in self.filtered(lambda p: not isinstance(p.id, models.NewId)):
+            link.name = (
+                "Link %s and %s" % (link.parent_id.name, link.child_id.name)
+                if link.parent_id and link.child_id
+                else "Unlink parent of %s" % link.child_id.name
+            )
+
+    @api.constrains("parent_id", "child_id")
+    def constrain_links(self):
+        """Check links against constriants.
+        These need to be checked against associated links as they may interact.
+        """
+        # Get all links that are associated with the same move and/or moveline:
+        move_lines = self.mapped("move_line_ids")  # Direct move lines
+        # Add the move lines of associated moves
+        move_lines |= move_lines.mapped("move_id").mapped("move_line_ids")
+        links = move_lines.mapped("u_result_package_link_ids")
+        # If some links are not associated with movelines, we still want to check them:
+        links |= self
+        links._validate_links()
+
+    def _validate_links(self):
+        """Validate package links to ensure that no constraints are broken."""
+        max_package_depth = self.env.user.get_user_warehouse().u_max_package_depth
+
+        # Sanitize links, check for repeated children as we should not have any
+        # Only exception may be moving a package from one package into another package. This will
+        # create a 'un-link' (no-parent) and a 'link' (with parent).
+        children_count = {}
+        for link in self:
+            children_count[link.child_id] = children_count.get(link.child_id, 0) + 1
+        for child_id, count in children_count.items():
+            if count == 2:
+                repeated_links = self.filtered(lambda x: x.child_id == child_id)
+                parent_ids = [link.parent_id.id for link in repeated_links]
```

Comment:
would this complain if we created the same link twice?

Reply:
Yes it would

Reply:
Should it?

Reply:
I can't think of a good reason we should have 2 identical links. It would either have to be multiple identical links on a single move line (which would be redundant and should not be created in the first place) or multiple move lines with the same links which would imply that the same stock transfer has been requested multiple times.

Reply:
Fair enough